### PR TITLE
[DNM] Wip rbd types

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

**Comment by wjwithagen** [comment]
@tchaikov 
Saw the DNM too late, but building gives:
```
usr/bin/ld: error: undefined symbol: rbd::mirror::image_map::PolicyMetaNone::TYPE
>>> referenced by Types.cc:30 (/home/jenkins/workspace/ceph-master/src/tools/rbd_mirror/image_map/Types.cc:30)
>>>               Types.cc.o:(void rbd::mirror::image_map::(anonymous namespace)::EncodeVisitor::operator()<rbd::mirror::image_map::PolicyMetaNone>(rbd::mirror::image_map::PolicyMetaNone const&) const) in archive ../../../lib/librbd_mirror_types.a

/usr/bin/ld: error: undefined symbol: rbd::mirror::image_map::PolicyMetaUnknown::TYPE
>>> referenced by Types.cc:30 (/home/jenkins/workspace/ceph-master/src/tools/rbd_mirror/image_map/Types.cc:30)
>>>               Types.cc.o:(void rbd::mirror::image_map::(anonymous namespace)::EncodeVisitor::operator()<rbd::mirror::image_map::PolicyMetaUnknown>(rbd::mirror::image_map::PolicyMetaUnknown const&) const) in archive ../../../lib/librbd_mirror_types.a
c++: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by tchaikov** [comment]
```[ RUN      ] TestMockImageMap.SetLocalImages
seed 38267
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/tools/rbd_mirror/image_map/StateTransition.cc: In function 'static const rbd::mirror::image_map::StateTransition::Transition& rbd::mirror::image_map::StateTransition::transit(rbd::mirror::image_map::StateTransition::ActionType, rbd::mirror::image_map::StateTransition::State)' thread ffff665ecf40 time 2017-11-16 04:39:19.409072
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/tools/rbd_mirror/image_map/StateTransition.cc: 78: FAILED assert(false)
 ceph version 13.0.0-3211-g38574bd (38574bd4b5618468d7e03a38e5d5a2098ebb638d) mimic (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0xfc) [0xffff9906b64c]
 2: (()+0x7cefb8) [0xaaaae6c20fb8]
 3: (rbd::mirror::image_map::Policy::start_next_action(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xb4) [0xaaaae6c1ba2c]
 4: (()+0x4b9a18) [0xaaaae690ba18]
 5: (FunctionContext::finish(int)+0x3c) [0xaaaae6923bb4]
 6: (Context::complete(int)+0x1c) [0xaaaae691b4ec]
 7: (()+0x4b5098) [0xaaaae6907098]
 8: (FunctionContext::finish(int)+0x3c) [0xaaaae6923bb4]
 9: (Context::complete(int)+0x1c) [0xaaaae691b4ec]
 10: (ThreadPool::worker(ThreadPool::WorkThread*)+0xa44) [0xffff99072c04]
 11: (ThreadPool::WorkThread::entry()+0x14) [0xffff99073b74]
 12: (()+0x6fc4) [0xffffa1982fc4]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
```

same old error.

File: src/tools/rbd_mirror/image_map/Types.cc [hunk 1]
```diff
@@ -28,7 +28,7 @@ class EncodeVisitor : public boost::static_visitor<void> {
 
   template <typename T>
   inline void operator()(const T& t) const {
-    ::encode(static_cast<uint32_t>(T::TYPE), m_bl);
+    ::encode(T::TYPE, m_bl);
```
[the remark this answers is not in the document]
Comment (dillaman) [reply src/tools/rbd_mirror/image_map/Types.cc]: https://github.com/ceph/ceph/blob/master/src/librbd/journal/Types.cc#L31
https://github.com/ceph/ceph/blob/master/src/librbd/watcher/Utils.h#L46
...

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/tools/rbd_mirror/image_map/Types.cc]: so it's already fixed? as `sizeof(unsigned) == sizeof(uint32_t)`.

[the remark this answers is not in the document]
Comment (dillaman) [reply src/tools/rbd_mirror/image_map/Types.cc]: The static cast should take care of any size deltas (although I'd expect sizeof(unsigned) to equal sizeof(uint32_t) unless you are on a 16bit CPU).

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/tools/rbd_mirror/image_map/Types.cc]: okay. that's also my understanding.

**Pull request closed without merging**